### PR TITLE
f-score + Tensorboard implementation

### PR DESCRIPTION
**Pull request by teodora99**

### CODE REVIEW

File: integration_tests/data_pipeline/training_entry_test.py [hunk 1]
```diff
@@ -7,26 +7,27 @@
...
 class TrainingEntryTest(absltest.TestCase):
 
-  def test_training_entry_amr_with_i(self):
-    sentence = 'Am I being foolish in doing this ?'
-    sentence = sentence.split()
-    amr_str = """
+    def test_training_entry_amr_with_i(self):
+        sentence = 'Am I being foolish in doing this ?'
+        sentence = sentence.split()
+        amr_str = """
```
Comment (Ranapop) [inline integration_tests/data_pipeline/training_entry_test.py]: why did you increase identation?

File: models.py [hunk 2]
```diff
@@ -11,20 +10,24 @@
...
 
-  def forward(self, inputs: torch.Tensor, input_lengths: torch.Tensor):
-    """
+    def __init__(self, input_vocab_size, hidden_size, use_bilstm=False):
+        super(Encoder, self).__init__()
+        self.embedding = nn.Embedding(input_vocab_size, EMB_DIM)
+        # dropout – If non-zero, introduces a Dropout layer on the outputs of each LSTM layer except the last layer,
+        # with dropout probability equal to dropout. Default: 0
+        # self.lstm = nn.LSTM(
+        #   EMB_DIM, hidden_size, NUM_LAYERS, dropout=0.5, bidirectional=use_bilstm)
```
Comment (Ranapop) [inline models.py]: remove comments

**Review by Ranapop — CHANGES_REQUESTED** [review]
General comment: The formatting was changed in a lot of files (indentation increased). This makes it appear like there are more changes than there actually are + it makes the comments pop out (they seem badly formatted).
Another issue is adding comments for dropout/commenting out snippets of code. You can put short TODOs but don't comment out large chunks that are the start of an implementation that should be done on a different branch.

File: train_concept_identification.py [hunk 3]
```diff
@@ -41,6 +42,90 @@ def compute_loss(criterion, logits, gold_outputs):
...
+        recall = true_positive / (true_positive + false_negative)
+    f_score = 0
+
+    if precision + recall != 0:
+        f_score = 2 * (precision * recall) / (precision + recall)
+
+    return f_score
+
+
+def teasor_to_list(gold_outputs, predicted_outputs):
```
Comment (Ranapop) [inline train_concept_identification.py]: typo: tensor_to_list

File: train_concept_identification.py [hunk 4]
```diff
@@ -41,6 +42,90 @@ def compute_loss(criterion, logits, gold_outputs):
...
+                break
+            else:
+                sentence_no_padding.append(word)
+        gold_list_no_padding.append(sentence_no_padding)
+
+    predicted_list_no_padding = []
+    for sentence in predicted_list_with_padding:
+        sentence_no_padding = []
+        for word in sentence:
+            if word == 1:
```
Comment (Ranapop) [inline train_concept_identification.py]: why stop at 1? who is 1?

File: train_concept_identification.py [hunk 5]
```diff
@@ -41,6 +42,90 @@ def compute_loss(criterion, logits, gold_outputs):
...
+    f_score
+  """
+
+    predicted_outputs_no_padding, gold_outputs_no_padding = teasor_to_list(gold_outputs, predicted_outputs)
+
+    ids_to_concepts_list_gold = list(vocabs.concept_vocab.keys())
+    concepts_as_list_gold = []
+    for sentence in gold_outputs_no_padding:
+        concepts_as_list_gold = [ids_to_concepts_list_gold[id] for id in sentence if
+                                 ids_to_concepts_list_gold[id] != UNK]
```
Comment (Ranapop) [inline train_concept_identification.py]: I'm not so sure about this. This could artificially increase the dev score. Maybe the initial gold sequence (before numericalization) should be stored instead & used here.
It could be done in a different branch, but please leave at least a TODO here

File: train_concept_identification.py [hunk 6]
```diff
@@ -41,6 +42,90 @@ def compute_loss(criterion, logits, gold_outputs):
   loss = criterion(flattened_logits, flattened_gold_outputs)
   return loss
 
+
+def compute_fScore(gold_outputs, predicted_outputs):
```
Comment (Ranapop) [inline train_concept_identification.py]: this method should have a unit test

**Review by Ranapop — CHANGES_REQUESTED** [review]

File: train_concept_identification.py (hunk 4 above)
Comment (teodora99) [reply train_concept_identification.py]: 1 is EOS.

File: train_concept_identification.py [hunk 7]
```diff
@@ -41,6 +42,94 @@ def compute_loss(criterion, logits, gold_outputs):
...
+            if ids_to_concepts_list_gold[id] != UNK:
+                concepts_as_list_gold.append(ids_to_concepts_list_gold[id])
+
+
+    ids_to_concepts_list_predicted = list(vocabs.concept_vocab.keys())
+    concepts_as_list_predicted = []
+    for sentence in predicted_outputs_no_padding:
+        for id in sentence:
+            if ids_to_concepts_list_predicted[int(id)] != UNK:
+                concepts_as_list_predicted.append(ids_to_concepts_list_predicted[int(id)])
```
Comment (Ranapop) [inline train_concept_identification.py]: This code is a repetition of the code a few lines above. Please extract in a function
It can be named remove_unk / remove_unk_from_sequence or something along those lines

File: train_concept_identification.py [hunk 8]
```diff
@@ -41,6 +42,94 @@ def compute_loss(criterion, logits, gold_outputs):
...
+
+    predicted_list_no_padding = []
+    for sentence in predicted_list_with_padding:
+        sentence_no_padding = []
+        for word in sentence:
+            if int(word) == 1:
+                break
+            else:
+                sentence_no_padding.append(word)
+        predicted_list_no_padding.append(sentence_no_padding)
```
Comment (Ranapop) [inline train_concept_identification.py]: This code repeats ~the same thing for the predicted and gold sequence. Please extract the common functionality in a function and call it for each sequence

**Review by Ranapop — COMMENTED** [review]
There's some code duplication that should be addressed -- see comments

File: train_concept_identification.py (hunk 4 above)
Comment (Ranapop) [reply train_concept_identification.py]: Yes, but you are relying on the fact that EOS was put first into the vocab after PAD -- not to mention it's a magic number in code. Please send the index for EOS; you should be able to get that using the vocab and the `EOS` constant imported from `data_pipeline.dataset`

File: train_concept_identification.py [hunk 9]
```diff
@@ -41,6 +42,91 @@ def compute_loss(criterion, logits, gold_outputs):
...
+                break
+            else:
+                sentence_no_padding.append(word)
+        list_no_padding.append(sentence_no_padding)
+    return list_no_padding
+
+
+def remove_unk_from_sequence(outputs_no_padding,
+                             vocabs: Vocabs):
+    # TODO store the gold data before numericalization and use it here
```
Comment (Ranapop) [inline train_concept_identification.py]: move the comment before calling the function on the gold data

File: train_concept_identification.py [hunk 10]
```diff
@@ -41,6 +42,91 @@ def compute_loss(criterion, logits, gold_outputs):
...
+    ids_to_concepts_list = list(vocabs.concept_vocab.keys())
```
Comment (Ranapop) [inline train_concept_identification.py:121]: Either move this out of the function or rename the function
At this point the function converts indices to strings and removes UNK tokens
It's an unexpected behaviour to have a function called 'remove_unk_from_sequence' also do the denumericalization -- the other way around might be better
Either rename to denumericalize/indices_to_words/indinces_to_strings or smth along those lines or transform the indices to words before this call and only have this call remove the UNK tokens

File: train_concept_identification.py [hunk 11]
```diff
@@ -41,6 +42,91 @@ def compute_loss(criterion, logits, gold_outputs):
...
+    concepts_as_list = []
+    for sentence in outputs_no_padding:
+        for id in sentence:
+            if ids_to_concepts_list[int(id)] != UNK:
+                concepts_as_list.append(ids_to_concepts_list[int(id)])
```
Comment (Ranapop) [inline train_concept_identification.py]: this can be done with list comprehension but not mandatory

**Review by Ranapop — CHANGES_REQUESTED** [review]

**Review by Ranapop — APPROVED** [review]
Overall looks good - I might have some extra comments but it would be nitpicking. Good job!
